### PR TITLE
Add provided methods `Seek::{stream_len, stream_position}`

### PR DESCRIPTION
**Pull request by LukasKalbertodt**
**Edit**: [Tracking issue](https://github.com/rust-lang/rust/issues/59359)

---

This adds two new, provided methods to the `io::Seek` trait:
- `fn stream_len(&mut self) -> Result<u64>`
- `fn stream_position(&mut self) -> Result<u64>`

Both are added for convenience and to improve readability in user code. Reading `file.stream_len()` is much better than to manually seek two or three times. Similarly, `file.stream_position()` is much more clear than `file.seek(SeekFrom::Current(0))`.

You can find prior discussions [in this internals thread](https://internals.rust-lang.org/t/pre-rfc-idea-extend-io-seek-with-convenience-methods-with-e-g-stream-len/9262). I think I addressed all concerns in that thread.

I already wrote three RFCs to add a small new API to libstd but I noticed that many public changes to libstd happen without an RFC. So I figured I can try opening a PR directly without going through RFCs first. After all, we do have rfcbot here too. If you think this change is too big to merge without an RFC, I can still close this PR and write an RFC.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Seems plausible to me! Since this is changing a pretty core trait, even though these functions are unstable, I'd want to gain some other feedback too:

@rfcbot fcp merge

**Comment by dtolnay** [comment]
`current_position` looks very similar to [`Cursor::position`](https://doc.rust-lang.org/std/io/struct.Cursor.html#method.position). Could we remove "current" from the name or does that make it less clear?

@rfcbot concern position

**Comment by dtolnay** [comment]
Could `stream_len` be `len`? Would there be any Seek impl that has a "stream len" different from its "len" or for which "len" isn't an appropriate term?

@rfcbot concern len

**Comment by LukasKalbertodt** [comment]
I would be fine with changing both names!

The only (very minor) argument against `len` I can think of: `len` has a precedence in many data structures where the method just returns an already stored value. For example, slices, vectors and everything in `std::collections` just return a stored length in `len()`. As such, people might assume `Seek::len` is also just a simple getter. But I don't really think it's a strong argument.

**Comment by sfackler** [comment]
This len would return a result, which should indicate that there's something bigger going on.

**Comment by jonas-schievink** [comment]
The ICE will be fixed once a new beta is cut that includes https://github.com/rust-lang/rust/pull/58501, but the bigger issue is that `tempfile` attempts to call `Cursor::position`, but for some reason ends up calling the new `Seek::position` (inherent impls should be preferred over trait impls, though?)

**Comment by LukasKalbertodt** [comment]
Interesting! I just pushed the renaming changes (`len` and `position`). I already thought that the same name on `Cursor` and `Seek` might lead to problems, but I wasn't sure. Thankfully Travis showed it to us.

First I was confused, but then I remembered [this question of dtolnay's Rust quiz](https://dtolnay.github.io/rust-quiz/23). Here is a [minimal example on Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d07c04b1e9c016d543e60319435fccf0).

So: keep `current_position` as name? I think that we are technically allowed to make these changes, because programmers could have prevented this by using a more explicit function call syntax. But I doubt that we practically want to make a change that breaks real crates out there. So what is the best way forward?

**Comment by jonas-schievink** [comment]
FWIW I also think that having a method `len()` that returns a `Result` is more confusing than naming it differently. I'm pretty sure it would be the first `len()` method in std that *isn't* just a getter (even `Metadata::len` is just a getter that returns `u64` directly).

**Comment by dtolnay** [comment]
Oh man, that quiz question. I think that rules out `position`.

I am mildly opposed to `current_position` because every getter is implicitly current -- we don't say HashMap::current_len and String::current_capacity for example. I would make an exception for getters for quantities that can be changed from outside of the current thread of execution. These are more like "measurements" than getters, and it may make sense to indicate that the current value of the quantity is measured but if you measure again you may get a different number. This applies to stream len (so if someone wanted `current_len` that would be justifiable) but not to position which only changes if we act on the stream from the current thread of execution.

I think we can land this with any name other than `position` and `current_position`. `io_position` or `stream_position` would be fine with me but open to other suggestions.

**Comment by dtolnay** [comment]
We can reconsider `len` also if people find `&mut self -> Result<u64>` to be a weird signature for something called len (though I don't expect it to cause problems). Bonus points for harmony between the len name and the position name, so that could be `io_len` + `io_position` or `stream_len` + `stream_position`.

**Comment by LukasKalbertodt** [comment]
> I think that rules out `position`.

Is that really the case, though? [RFC 1105](https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md#minor-change-adding-a-defaulted-item) defines what changes are considered "major" and which ones are considered "minor". It says: 

> ### Minor change: adding a defaulted item
>
> Adding a defaulted item is technically a breaking change: [...] According to the basic principles of this RFC, such a change is minor: it is always possible to annotate the call [...].

The RFC uses a different example to show that adding a defaulted item can be a breaking change, but I guess the reasoning still applies.

Now of course we wouldn't want to have a change that we know causes real breakage in the world -- regardless if it's minor or major according to the RFC. So I guess what I'm asking is: do you consider this one breakage in the compiler suite as enough evidence that such a change is not OK? Or would it be better to have a crater run? 

As @jonas-schievink pointed out in a private chat with me, [the code that breaks in Travis](https://github.com/Stebalien/tempfile/blob/251a76a3c5e76cfcd90ea986764f6a5ab2d4e1be/src/spooled.rs#L88-L92) is extremely fishy: neither `get_ref` nor `position` of `Cursor` need a mutable reference, so `ref cursor` should be enough. Furthermore the mutable reference is then explicitly dropped which shouldn't do anything because it's at the end of its scope anyway. That doesn't mean that there can't be non-fishy code that breaks under this change to `Seek`, but I guess it's still worth mentioning that in that particular case, everything would be fine if the code was "more correct".

---

> but not to position which only changes if we act on the stream from the current thread of execution.

That's not necessarily true, right? Someone could implement `Seek` for a type that can be shared across multiple threads. (Although I still agree that `current_` in `current_position` is kinda bad.)

I don't like the `io_` prefix. I would be fine with `stream_len` and `stream_position` (although I'm not super happy about the `stream_position`). Alternatives: `seek_position` or `position_in_stream`?

**Comment by LukasKalbertodt** [comment]
I renamed the methods again to: `stream_len` and `stream_position` (one of @dtolnay's suggestions). This should avoid any name-clashes and make this PR mergable. Figuring out the exact name can probably be done in the tracking issue after merging this, right?

**Comment by dtolnay** [comment]
Thanks, that works for me.

@rfcbot resolve position
@rfcbot resolve len

**Comment by Dr-Emann** [comment]
Are there plans to specialize this where possible?

* `Cursor` can trivially get the stream length without seeking twice (might get optimized out anyway?)
* `File` can try to go through metadata.len, before trying to fall back to the double-seek?

**Comment by Dr-Emann** [comment]
Are we also allowed to do something like

```rust
let current = cursor.seek(SeekFrom::Current(0))?;
let end = cursor.seek(SeekFrom::End(0))?;
if (current != end) {
    cursor.seek(SeekFrom::Start(current))?;
}
```
To prevent the third call to seek if we are at the end already?

**Comment by hellow554** [comment]
>     * `File` can try to go through metadata.len, before trying to fall back to the double-seek?

But then you have to sync the metadata first, which can be more expensive than two seeks (see [Stackoverflow](https://stackoverflow.com/questions/54303398/how-to-get-the-size-of-an-already-opened-file-in-rust/54303584#54303584))

File: src/libstd/io/mod.rs [hunk 1]
```diff
@@ -1332,10 +1332,11 @@ pub trait Seek {
 
     /// Returns the length of this stream (in bytes).
     ///
-    /// This method is implemented using three seek operations. If this method
-    /// returns successfully, the seek position is unchanged (i.e. the position
-    /// before calling this method is the same as afterwards). However, if this
-    /// method returns an error, the seek position is undefined.
+    /// This method is implemented using uo to three seek operations. If this
```
Comment (hellow554) [inline src/libstd/io/mod.rs]: uo?

File: src/libstd/io/mod.rs [hunk 2]
```diff
@@ -1332,10 +1332,11 @@ pub trait Seek {
...
+    /// method returns successfully, the seek position is unchanged (i.e. the
+    /// position before calling this method is the same as afterwards).
+    /// However, if this method returns an error, the seek position is
+    /// undefined.
```
Comment (hellow554) [inline src/libstd/io/mod.rs]: I thought safe rust has no undefined behavoir? Is this UB anyway?

File: src/libstd/io/mod.rs (hunk 1 above)
Comment (LukasKalbertodt) [reply src/libstd/io/mod.rs]: *sigh*

It's supposed to be "up". Sorry for that and thanks for catching the typo. Gonna fix in a minute...

File: src/libstd/io/mod.rs (hunk 2 above)
Comment (LukasKalbertodt) [reply src/libstd/io/mod.rs]: Yes and no. The English meaning of "undefined behavior" is different from what most people mean by it. So what this function says is that the *position* is undefined. That's all, no other things can be undefined. People usually use UB to say "everything" is undefined -- including the contents of your `/root` directory and the state of your laundry.

I heard people use the term "unspecified behavior" to mean: "some things are not specified/undefined, but it's not the super bad UB".

I can change the word to "unspecified" if you think that is clearer?

Comment (hellow554) [reply src/libstd/io/mod.rs]: "the seek position is indetermined"?

Comment (trentj) [reply src/libstd/io/mod.rs]: "unspecified" is used for similar situations elsewhere in the standard library, e.g. [It is unspecified how many elements are removed from the vector if the `Drain` value is leaked.](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain)

Comment (LukasKalbertodt) [reply src/libstd/io/mod.rs]: I changed it to "unspecified" for now.

**Comment by Dr-Emann** [comment]
> > * `File` can try to go through metadata.len, before trying to fall back to the double-seek?
> 
> But then you have to sync the metadata first, which can be more expensive than two seeks (see [Stackoverflow](https://stackoverflow.com/questions/54303398/how-to-get-the-size-of-an-already-opened-file-in-rust/54303584#54303584))

Is that accurate? I thought `sync_all` is for ensuring changes make it to disk (in case of shutdown, etc), but metadata should still be consistent on a running system.

However, I did find another possible reason to prefer seeking, even on files: files in `/proc` are reported as 0 sized in file metadata, whereas trying to get the length by seeking fails

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=923ffea90a1e35d53c79ddc3a134619d

File: src/libstd/io/mod.rs [hunk 3]
```diff
@@ -1329,6 +1329,85 @@ pub trait Seek {
...
+    ///
+    /// This method is implemented using up to three seek operations. If this
+    /// method returns successfully, the seek position is unchanged (i.e. the
+    /// position before calling this method is the same as afterwards).
+    /// However, if this method returns an error, the seek position is
+    /// unspecified.
+    ///
+    /// If you need to obtain the length of *many* streams and you don't care
+    /// about the seek position afterwards, you can reduce the number of seek
+    /// operations by simply calling `seek(SeekFrom::End(0))` and use its
```
Comment (tbu-) [inline src/libstd/io/mod.rs]:
```suggestion
    /// operations by simply calling `seek(SeekFrom::End(0))` and using its
```

File: src/libstd/io/mod.rs [hunk 4]
```diff
@@ -1329,6 +1329,85 @@ pub trait Seek {
...
+    /// return value (it is also the stream length).
+    ///
+    /// Note that length of a stream can change over time (for example, when
+    /// data is appended to a file). So calling this method multiply times does
```
Comment (tbu-) [inline src/libstd/io/mod.rs]:
```suggestion
    /// data is appended to a file). So calling this method multiple times does
```

File: src/libstd/io/mod.rs (hunk 3 above)
Comment (LukasKalbertodt) [reply src/libstd/io/mod.rs]: Thanks for noticing and thanks for letting me try this new suggestion feature of GitHub. Very nice ^_^

File: src/libstd/io/mod.rs [hunk 5]
```diff
@@ -1329,6 +1329,85 @@ pub trait Seek {
...
+    ///
+    /// fn main() -> io::Result<()> {
+    ///     let mut f = File::open("foo.txt")?;
+    ///
+    ///     let len = f.stream_len()?;
+    ///     println!("The file is currently {} bytes long", len);
+    ///     Ok(())
+    /// }
+    /// ```
+    #[unstable(feature = "seek_convenience", issue = "0")]
```
Comment (lzutao) [inline src/libstd/io/mod.rs]: So we don't have tracking issue for this?

Comment (LukasKalbertodt) [reply src/libstd/io/mod.rs]: Oopsie! There is a tracking issue now: https://github.com/rust-lang/rust/issues/59359
And I added the number to the code here: https://github.com/rust-lang/rust/pull/59360

Thanks for catching this.

**Pull request merged**